Protect against race conditions

amshinde opened the pull request.

amshinde committed:
sync: Synchronize container map in pod struct

Maps in golang are not safe for concurrent use.
Although the serial port that is used by the agent translates
to some degree of serialization, we should protect the containers
map from access by several go-routines.
There is a chance that removecontainer may be called while
a container is cleaning up its process pipes.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde committed:
agent: Refactor closeProcessStreams/Pipes to be called on container

The functions closeProcessStreams() and closeProcessPipes were being
called on pod struct pointer. Refactor these to be called on
container instead. Semantically it makes sense to have these
methods called on a container.
This also cleans up the code as we avoid accesses to container map
in the pod to fetch a container in order to access the container's
processes map.

The next step would be to protect the processes maps in container
struct for concurrent acceses as these are very much likely
to be accessed concurrently when several processes in a container
are exiting.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

sboeuf commented:
@amshinde I'll take a look at this today

sboeuf commented:
@amshinde actually I have noticed the `wip` in the title, let me know when it is ready so that I can review !

amshinde commented:
@sboeuf PR is ready for review

amshinde committed:
agent: Synchronize the read/writes to processes in container struct

The processes map needs be protected against concurrent read and
writes. Several exec'ed processes for a container can update the
processes map of a container concurrently.

Fixes #76

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

devimc commented:
lgtm

[![Approved with PullApprove](https://img.shields.io/badge/approvers-approved-brightgreen.svg)](https://pullapprove.com/clearcontainers/agent/pull-request/80/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent) [![Approved with PullApprove](https://img.shields.io/badge/reviewers-approved-brightgreen.svg)](https://pullapprove.com/clearcontainers/agent/pull-request/80/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

amshinde commented:
Looks like a couple of tests are timing out while killing a container in semaphore. Will be able to take a look at this tomorrow.
@sboeuf  let me know what you think about the changes.

sboeuf commented on agent.go:
```diff
@@ -120,15 +120,16 @@ type container struct {
+	network        hyper.Network
+	stdinLock      sync.Mutex
+	ttyLock        sync.Mutex
+	containersLock sync.RWMutex
```
I would call this one `podLock` instead of `containersLock` since it actually locks the Pod.

amshinde replied:
@sboeuf This only protects the containers map, so I have used the name containersLock here.

sboeuf replied:
Fair enough.

sboeuf commented on agent.go:
```diff
@@ -197,6 +198,24 @@ func main() {
+	p.containers[ID] = ctr
+	p.containersLock.Unlock()
+}
+
```
You should create a new function `deleteContainer()` locking the removal of a container:
```go
func (p *pod) deleteContainer(id string) {
    p.containersLock.Lock()
    delete(p.containers, id)
    p.containersLock.Unlock()
}
```
And use it from the code below

sboeuf replied:
Great, this has been added.

sboeuf commented on agent.go:
```diff
@@ -1096,11 +1124,13 @@ func removeContainerCb(pod *pod, data []byte) error {
 		return err
 	}
 
+	pod.containersLock.Lock()
```
This won't be enough here. Mutexes are a first step, but we also needs some synchronization waiting for all the started processes to end before we try to remove the container. Basically, this means we need to wait on a `sync` before to perform the container's removal here. Every time we create/start a new process, we should do something like `wg.Add(1)` and every time it ends, call into `wg.Done()`. This way we ensure we are not removing a container which is still used.

amshinde replied:
done

sboeuf commented on agent.go:
```diff
@@ -117,6 +117,7 @@ type container struct {
 	container libcontainer.Container
 	config    configs.Config
 	processes map[string]*process
+	pod       *pod
```
Refactoring sounds okay to me, but you should give the pod pointer through the parameters of the function `closeProcessStreams()` since you need the pod pointer only in this function. This would avoid to add this pointer in every container.

Moreover, please fix your commit message, there are some syntax/grammar errors.

sboeuf replied:
This comment has not been addressed.

sboeuf commented on agent.go:
```diff
@@ -114,10 +114,11 @@ type process struct {
+	config        configs.Config
+	processes     map[string]*process
+	pod           *pod
+	processesLock sync.RWMutex
```
I would call this one `containerLock` instead of `processesLock` since it locks the entire container.

amshinde replied:
@sboeuf This only protects the processes map, hence the name.

sboeuf replied:
Fair enough.

sboeuf commented on agent.go:
```diff
@@ -565,67 +566,99 @@ func (p *pod) sendSeq(seq uint64, data []byte) error {
+	c.processes[pid] = process
+	c.processesLock.Unlock()
+}
+
```
You could add the same type of function for the process removal:
```go
func (c *container) deleteProcess(pid string) {
    ...
}
```

sboeuf replied:
Great, you have addressed this one.

sboeuf commented on agent.go:
```diff
@@ -197,6 +198,24 @@ func main() {
 	wgLoops.Wait()
 }
 
+func (p *pod) getContainer(ID string) (ctr *container) {
```
`ID` should be lower case `id`.

sboeuf commented on agent.go:
```diff
@@ -197,6 +198,24 @@ func main() {
+	return ctr
+}
+
+func (p *pod) setContainer(ID string, ctr *container) {
```
`ID` should be lower case `id`.

amshinde committed:
sync: Wait for processes before removing container

Add a wait group for waiting on container processes.
Increment wait group counter for a container when a new
container is started or when a new process is exec'ed for
an existing container.
Wait for all processes to exit, before removing a
container from its pod.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde commented:
@sboeuf  I have pushed changes, have a look

sboeuf commented:
Thanks @amshinde

sboeuf commented on agent.go:
```diff
@@ -114,10 +114,11 @@ type process struct {
 }
 
 type container struct {
```
Please rework the commit message, there are some syntax/grammar errors.

sboeuf commented on agent.go:
```diff
@@ -1103,6 +1105,8 @@ func newContainerCb(pod *pod, data []byte) error {
 
 	pod.setContainer(payload.ID, container)
 
+	container.wgProcesses.Add(1)
```
This can be moved to `runContainerProcess()` in order to avoid calling this both from `newContainerCb()` and `execCb()`. There is no risk to increase the waitgroup counter from this go routine as long as we make sure the counter is incremented before the signal is sent through the `started` Go channel.

amshinde replied:
@sboeuf iirc Add should execute before the statement creating the goroutine and not in the goroutine,  hence the two places.

sboeuf replied:
Oh... You think it does not work even if we have the handler `container.wgProcesses` from the goroutine ?
I thought it would be okay since the starting of any process (newContainer or exec) waits for the `started` channel for synchronization.

sboeuf commented on agent.go:
```diff
@@ -1229,6 +1235,8 @@ func execCb(pod *pod, data []byte) error {
 
 	ctr.setProcess(payload.Process.ID, process)
 
+	ctr.wgProcesses.Add(1)
```
Ditto. (can be moved)

sboeuf reviewed (changes requested):
Few comments

amshinde commented:
@sboeuf Commit messages have been fixed

sboeuf commented:
LGTM
